File: packages/ts/src/infrastructure/store/redis-run-store.ts
import type { RunStore } from "../../application/ports/run-store";
import type { RunId } from "../../domain/identifiers";
import type { RunState } from "../../domain/run";

/** The subset of a Redis client the run store needs. */
export interface RedisStringCommands {
  get(key: string): Promise<string | null>;
  set(key: string, value: string): Promise<unknown>;
}

const KEY_PREFIX = "airlock:run:";

/** Persists run state in Redis as JSON (see ADR-0004). */
export class RedisRunStore implements RunStore {
  constructor(private readonly redis: RedisStringCommands) {}

  async save(state: RunState): Promise<void> {
    await this.redis.set(`${KEY_PREFIX}${state.runId}`, JSON.stringify(state));
  }

  async load(runId: RunId): Promise<RunState | null> {
    // The store only ever holds run state we serialized ourselves.
    const raw = await this.redis.get(`${KEY_PREFIX}${runId}`);
    return raw === null ? null : (JSON.parse(raw) as RunState);
  }
}

File: packages/py/src/airlock/infrastructure/store/redis_run_store.py
"""Redis-backed run store: run state persisted as JSON (ADR-0004)."""

from __future__ import annotations

from typing import TYPE_CHECKING

from airlock.domain.run import RunState

if TYPE_CHECKING:
    from redis.asyncio import Redis

    from airlock.domain.identifiers import RunId

_KEY_PREFIX = "airlock:run:"


class RedisRunStore:
    def __init__(self, client: Redis) -> None:
        self._client = client

    async def save(self, state: RunState) -> None:
        await self._client.set(f"{_KEY_PREFIX}{state.run_id}", state.model_dump_json())

    async def load(self, run_id: RunId) -> RunState | None:
        raw = await self._client.get(f"{_KEY_PREFIX}{run_id}")
        if raw is None:
            return None
        return RunState.model_validate_json(raw)
